### PR TITLE
Do not ICE on missing access place description during mutability error reporting

### LINKED ISSUE #61187: Internal compiler error in nightly when reversing a vector inside a async fn
Got an internal compiler error when reversing a vector inside a async fn

I tried this code:

```rust
#![feature(async_await)]

fn main() {
}

async fn response(data: Vec<u8>) {
    data.reverse();
}
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4e90f405e1d73fdbc7b009a3de23bcae))

This happened:

```
$ RUST_BACKTRACE=1 cargo check
    Checking rust-error-testing-unfold-option-unwrap v0.1.0 (/home/sindre/src/rust/rust-error-testing-unfold-option-unwrap)
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:347:21
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:478
   6: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   7: rust_begin_unwind
             at src/libstd/panicking.rs:308
   8: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   9: core::panicking::panic
             at src/libcore/panicking.rs:49
  10: rustc_mir::borrow_check::mutability_errors::<impl rustc_mir::borrow_check::MirBorrowckCtxt>::report_mutability_error
  11: rustc_mir::borrow_check::MirBorrowckCtxt::access_place
  12: <rustc_mir::borrow_check::MirBorrowckCtxt as rustc_mir::dataflow::DataflowResultsConsumer>::visit_statement_entry
  13: rustc_mir::borrow_check::do_mir_borrowck
  14: rustc::ty::context::GlobalCtxt::enter_local
  15: rustc_mir::borrow_check::mir_borrowck
  16: rustc::ty::query::__query_compute::mir_borrowck
  17: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::mir_borrowck>::compute
  18: rustc::dep_graph::graph::DepGraph::with_task_impl
  19: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  20: rustc::ty::<impl rustc::ty::context::TyCtxt>::par_body_owners
  21: rustc::util::common::time
  22: rustc_interface::passes::analysis
  23: rustc::ty::query::__query_compute::analysis
  24: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::analysis>::compute
  25: rustc::dep_graph::graph::DepGraph::with_task_impl
  26: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  27: rustc::ty::context::tls::enter_global
  28: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  29: rustc_interface::passes::create_global_ctxt::{{closure}}
  30: rustc_interface::interface::run_compiler_in_existing_thread_pool
  31: std::thread::local::LocalKey<T>::with
  32: scoped_tls::ScopedKey<T>::set
  33: syntax::with_globals
query stack during panic:
#0 [mir_borrowck] processing `response::{{closure}}#0`
#1 [analysis] running analysis passes on this crate
end of query stack

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.36.0-nightly (dec4c5201 2019-05-24) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `rust-error-testing-unfold-option-unwrap`.

To learn more, run the command again with --verbose.
```

## Meta

```
$ rustc --version --verbose
rustc 1.36.0-nightly (dec4c5201 2019-05-24)
binary: rustc
commit-hash: dec4c5201f88efbc3020b04ba96a5ee2c3b6cfcd
commit-date: 2019-05-24
host: x86_64-unknown-linux-gnu
release: 1.36.0-nightly
LLVM version: 8.0
```

### PR DESCRIPTION
**Pull request by estebank**
Fix #61187.

### CODE REVIEW

**Comment by estebank** [comment]
Cc @rust-lang/wg-diagnostics @cramertj for wording

**Comment by davidtwco** [comment]
I've only had a cursory glance, but would we be able to give the name of the place if we allowed `CompilerDesugaringKind::Async` here or would that break some cases?

https://github.com/rust-lang/rust/blob/46805805abe58c287fa16963f897fd09d5d97467/src/librustc_mir/borrow_check/error_reporting.rs#L283

**Comment by estebank** [comment]
We could probably detect the `Async` desugaring and say something like "cannot borrow async argument as mutable, as it is not declared as mutable"

File: src/test/ui/issues/issue-61187.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// edition:2018
```
Comment (Centril) [inline src/test/ui/issues/issue-61187.rs]: Move this test into the async/await folder (see https://github.com/rust-lang/rust/pull/61138).

**Comment by estebank** [comment]
@davidtwco I think it is a cool idea to use the span desugaring mark for this, but sadly at the moment we don't mark the argument's spans with it and it is taking me some time to change it to do so.

File: src/test/ui/issues/issue-61187.rs (hunk 1 above)
Comment (estebank) [reply src/test/ui/issues/issue-61187.rs]: Done

File: src/librustc_mir/borrow_check/mutability_errors.rs [hunk 2]
```diff
@@ -41,14 +41,19 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
...
-        let item_msg;
         let reason;
         let access_place_desc = self.describe_place(access_place);
         debug!("report_mutability_error: access_place_desc={:?}", access_place_desc);
 
+        let mut item_msg = match (&access_place_desc, &the_place_err) {
+            (Some(desc), _) => format!("`{}`", desc),
+            (None, Place::Base(PlaceBase::Local(local))) if self.mir.local_decls[*local]
+                .source_info.span.is_compiler_desugaring(CompilerDesugaringKind::Async)
+            => "async `fn` parameter".to_string(),
```
Comment (davidtwco) [inline src/librustc_mir/borrow_check/mutability_errors.rs]:
```suggestion
            => "`async fn` parameter".to_string(),
```

I can’t remember if there’s precedent for how we quote `async fn`.
(Somehow managed to accidentally comment this in the commit first, oops)

Comment (estebank) [reply src/librustc_mir/borrow_check/mutability_errors.rs]: Done

File: src/librustc_mir/borrow_check/mutability_errors.rs [hunk 3]
```diff
@@ -41,14 +41,19 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
...
         let reason;
         let access_place_desc = self.describe_place(access_place);
         debug!("report_mutability_error: access_place_desc={:?}", access_place_desc);
 
+        let mut item_msg = match (&access_place_desc, &the_place_err) {
+            (Some(desc), _) => format!("`{}`", desc),
+            (None, Place::Base(PlaceBase::Local(local))) if self.mir.local_decls[*local]
+                .source_info.span.is_compiler_desugaring(CompilerDesugaringKind::Async)
+            => "`async fn` parameter".to_string(),
+            (None, _) => "temporary place".to_string(),
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/mutability_errors.rs]: I don't see any output actually producing this. Should we `bug!` or at least `delay_span_bug` here?

**Comment by oli-obk** [comment]
@bors r+ p=5 async await priorization

my nit may be fixed in a separate PR

**Comment by Centril** [comment]
@bors p=0 rollup

That was slightly too over-prioritized 😂 -- including it in a rollup will do.

r? @oli-obk

**Comment by oli-obk** [comment]
"just following orders" :stuck_out_tongue:

**Comment by matthewjasper** [comment]
I don't think this is really the correct fix, the variable being mutated here shouldn't have it's Span marked. I guess this is fine for now.

**Comment by oli-obk** [comment]
The problem isn't that the argument has its `Span` marked, what we're seeing is a `temp` local that is part of the expansion of `async fn` causing the error. That temp local is directly connected to the argument where the `Span` points to.

**Comment by estebank** [comment]
> we should probably not mark the local's span during HIR lowering.

@oli-obk why shouldn't we? I believe there are a couple of places where we mark spans that aren't _strictly_ part of the desugared output, but that not marking would make it impossible to provide good diagnostics in some cases. I believe this might just be one of these cases :-/

**Comment by davidtwco** [comment]
In this instance, the span is marked to avoid `unused_mut` lint (since the desugaring will just pull the argument directly into the closure). We can’t remove the `mut` as that would affect the input that macros get. I’m hoping to tidy up the logic around that desugaring soon (because it’s unfortunately a mess) but I’m not sure if this span being marked will change with that.

**Comment by oli-obk** [comment]
Hmm... in that case, would it make sense to special case `append_local_to_string` instead of special casing on the output of it?

**Comment by matthewjasper** [comment]
The problems is that we are marking more than we should be. The lowered code here looks like:

```rust
async fn response(
    data: Vec<u8> // This is an internal variable, it should have its Span marked
) -> ::std::future::from_generator(move || {
    let data // This variable is effectively the user's parameter, it should not have it's Span marked
        = data; 
    data.reverse();
}) 
```

This is causing false negatives in cases like:

```rust
#![deny(unused_mut)]
pub async fn response(mut data: Vec<u8>) {
    let _ = data;
}
```

**Comment by davidtwco** [comment]
@estebank, I have a drive-by fix for this in #61413 which re-implements the async fn drop order lowering (sorry!).

**Pull request closed without merging**